### PR TITLE
botocore: always use x-ray for http header injection

### PR DESCRIPTION
**Pull request by tsloughter**
Updates for spec change https://github.com/open-telemetry/opentelemetry-specification/pull/3212

# Description

AWS will only propagate the x-ray header's context through to other services. Injecting context to any other header with a different propagator will result in the context being dropped.

The propagator is overridable by the user as an argument to BotocoreInstrumentor to give them final control.

## Type of change

Please delete options that are not relevant.

- [X] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [X] add `test_override_xray_propagator_injects_into_request`

# Does This PR Require a Core Repo Change?

- [ ] Yes. - Link to PR: 
- [X] No.

# Checklist:

See [contributing.md](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/CONTRIBUTING.md) for styleguide, changelog guidelines, and more.

- [X] Followed the style guidelines of this project
- [X] Changelogs have been updated
- [X] Unit tests have been added
- [ ] Documentation has been updated

### CODE REVIEW

File: instrumentation/opentelemetry-instrumentation-botocore/src/opentelemetry/instrumentation/botocore/__init__.py [hunk 1]
```diff
@@ -140,6 +135,10 @@ def _instrument(self, **kwargs):
         self.request_hook = kwargs.get("request_hook")
         self.response_hook = kwargs.get("response_hook")
 
+        propagator = kwargs.get("propagator")
+        if propagator != None:
```
Comment (ocelotl) [inline instrumentation/opentelemetry-instrumentation-botocore/src/opentelemetry/instrumentation/botocore/__init__.py:139]:
```suggestion
        if propagator is not None:
```

File: instrumentation/opentelemetry-instrumentation-botocore/src/opentelemetry/instrumentation/botocore/__init__.py [hunk 2]
```diff
@@ -149,13 +148,26 @@ def _instrument(self, **kwargs):
...
+            self._patched_endpoint_prepare_request,
         )
 
     def _uninstrument(self, **kwargs):
         unwrap(BaseClient, "_make_api_call")
         unwrap(Endpoint, "prepare_request")
 
+    # pylint: disable=unused-argument
+    def _patched_endpoint_prepare_request(
+        self, wrapped, instance, args, kwargs
```
Comment (ocelotl) [inline instrumentation/opentelemetry-instrumentation-botocore/src/opentelemetry/instrumentation/botocore/__init__.py:160]:
```suggestion
        self, wrapped, instance, *args, **kwargs
```

**Review by ocelotl — CHANGES_REQUESTED** [review]
Just minor changes requested.

**Review by ocelotl — APPROVED** [review]

**Comment by tsloughter** [comment]
So the failures are related to the `prepare_request` wrapper and use of `headers`:

```
>       headers = request.headers
E       AttributeError: 'tuple' object has no attribute 'headers'
```

Thing is.. before the patch it also used `request.headers` from `args[0]`. So it is maybe something with how I moved `_patched_endpoint_prepare_request` into `BotocoreInstrumentor`?

File: instrumentation/opentelemetry-instrumentation-botocore/src/opentelemetry/instrumentation/botocore/__init__.py [hunk 3]
```diff
@@ -149,13 +148,26 @@ def _instrument(self, **kwargs):
...
+            self._patched_endpoint_prepare_request,
         )
 
     def _uninstrument(self, **kwargs):
         unwrap(BaseClient, "_make_api_call")
         unwrap(Endpoint, "prepare_request")
 
+    # pylint: disable=unused-argument
+    def _patched_endpoint_prepare_request(
+        self, wrapped, instance, *args, **kwargs
```
Comment (nstawski) [inline instrumentation/opentelemetry-instrumentation-botocore/src/opentelemetry/instrumentation/botocore/__init__.py:160]: The original function used `args, kwargs` instead of variable-length arguments. When you revert this change, all tests pass.

Comment (tsloughter) [reply instrumentation/opentelemetry-instrumentation-botocore/src/opentelemetry/instrumentation/botocore/__init__.py]: Thanks! Updated.

**Comment by tsloughter** [comment]
Just some linting issues I can fix up soon.

File: instrumentation/opentelemetry-instrumentation-botocore/tests/test_botocore_instrumentation.py [hunk 4]
```diff
@@ -306,20 +302,44 @@ def check_headers(**kwargs):
...
+
+        BotocoreInstrumentor().instrument()
+
+        ec2 = self._make_client("ec2")
+        ec2.meta.events.register_first(
+            "before-send.ec2.DescribeInstances", check_headers
+        )
+        ec2.describe_instances()
+
+        request_id = "fdcdcab1-ae5c-489e-9c33-4637c5dda355"
```
Comment (ocelotl) [inline instrumentation/opentelemetry-instrumentation-botocore/tests/test_botocore_instrumentation.py:338]:
```suggestion
```

This variable is not being used, having it there fails lint.

**Review by ocelotl — COMMENTED** [review]
Just remove a variable to make lint pass.

**Comment by tsloughter** [comment]
@ocelotl it won't let me trigger a rerun of https://github.com/open-telemetry/opentelemetry-python-contrib/actions/runs/4985069665/jobs/8924251061?pr=1741 are you able to?
